### DISCUSSION TITLE
pattern_ui

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#28] [Announcements]
# 事件结构与程序界面

https://lv.qizhen.xyz/pattern_ui

**Comment by Arita9527** [comment]
你好，为什么例子Dynamic Event Generation.vi中VI引用可以有mouse move？
问题1.图中VI引用在哪里有用？
问题2.图中的mouse move为什么我点开找不到？
![why](https://ibb.co/195q28f)

**Reply by Arita9527** [reply]
![vi refnum](https://i.postimg.cc/kMzC3QvQ/VI-refnum.png)

**Reply by ruanqizhen** [reply]
1. 你是问从哪里找到的吗？控件选板 “Modern->Refnum“ 下面有一个 VI Refnum，把它拖到程序框图上，就是 VI 引用常量。
2. 使用 ”Mouse Move“ 事件需要提供更具体的目标。VI太过笼统，至少要说明是 VI 前面板，程序框图，还是某个控件？比如按下面的程序实现你就可以看到了：
![](https://github.com/ruanqizhen/labview_book/blob/main/docs/images_2/z044.png?raw=true)

**Reply by Arita9527** [reply]
1.原来前面板的控件选板中元素也可以拖进去程序框图啊？  我还纳闷为啥程序框图中找不到阿。
2.”比如按下面的程序实现你就可以看到了：”----没有看到呢？
3.新问题是：
---我用了VI refnum，尝试连接Reg Event，但是里面没有mouse MOVE。
![](https://i.postimg.cc/nL8pp6pK/VI-refnum-mouse-move.png)
---但是打开LV 例子中查看VI refnum，它选择的是VI。为什么却有mouse move？（你可以打开LV的动态注册事件例子看看哦）
![](https://i.postimg.cc/66h3z5wk/VI-refnum-mouse-move.png)
然后我在它下面建立一个reg events来连接它，也没有阿。
![](https://i.postimg.cc/N07qYFNY/mouse-move2.png)

**Reply by ruanqizhen** [reply]
这是LabVIEW中常出现的一种情况。那个示例VI可能最初是在某个较早版本的LabVIEW下编写的，那时候可以直接使用VI的mouse move事件。但是后续的LabVIEW版本不建议这样使用了，所以你就无法找到它了。如果你一定要用，可以直接复制这段代码，但是既然LabVIEW不建议这样用，肯定也是觉得这样用不太好。
所以最好还是按照目前版本允许的方式来用：使用前面板上某个面板的 mouse move事件。

**Reply by Arita9527** [reply]
明白，我用了panne的mouse move事件，效果出来了。

**Comment by Arita9527** [comment]
对了，这个#1代表什么意思呢？
![](https://i.postimg.cc/bJyQVVD8/1.png)

**Reply by ruanqizhen** [reply]
#1就表示输入的VI，因为你的vi引用不是给具体的某个VI的，所以没办法显示具体的VI，只能用个序号来代替。

**Comment by siquandaydayup** [comment]
ruan老师好，用户自定义事件，字符串长度超过10，发出警告，while循环咋停止啊，能否给个示范

**Reply by ruanqizhen** [reply]
你看一下这个示例可以解决你的问题吗： https://github.com/ruanqizhen/labview_book/blob/main/code/%E5%9F%BA%E6%9C%AC%E7%A8%8B%E5%BA%8F%E7%BB%93%E6%9E%84/%E4%BA%8B%E4%BB%B6%E7%BB%93%E6%9E%84/%E7%94%A8%E6%88%B7%E8%87%AA%E5%AE%9A%E4%B9%89%E4%BA%8B%E4%BB%B6.vi

这本书配备了很多示例程序，都在 github 对应的 code 文件夹下。

**Comment by whchrome** [comment]
阮老师，您好，我在使用用户事件时，发现程序像下面这样写，无法产生Init这个用户事件：
![1](https://user-images.githubusercontent.com/115133257/203886662-17b4631c-b902-493d-9874-df1a6f1e051f.png)
后来发现，是“用户事件引用句柄”连线分叉的原因，程序可能先执行“产生用户事件”，后执行“注册事件”，如果给“产生用户事件”前增加一个延时，就可以解决，如下图：
![2](https://user-images.githubusercontent.com/115133257/203886881-a1b15638-60d3-459b-a0d6-5145ae31064a.png)
但是我看到您文章中的程序，“注册事件”的“用户事件”右端有“用户事件引用句柄”的接口，基于LabVIEW的数据流决定的运行先后顺序，可以解决这个问题，如下图：
![4](https://user-images.githubusercontent.com/115133257/203887237-0fdad59d-f677-44f0-b927-77333324592c.png)
但是我在我的LabVIEW中使用“用户事件”，发现其右端没有“用户事件引用句柄”的接口，如下图，这是为什么，是新版本的功能吗？我用的是LabVIEW2020。
![3](https://user-images.githubusercontent.com/115133257/203887422-55319db3-9532-4161-8366-0aefe0ccb2df.png)

**Reply by ruanqizhen** [reply]
这是视觉的错觉 :)  注册事件右端没有这个输出，那个数据线就是输入的“用户事件”数据线，跟你的程序是一样的。真正决定运行顺序的是“错误”（error in/out）数据线。LabVIEW 中几乎所有复杂的函数都有 error in/out，可以用来相互连线，所以错误数据线除了做错误处理，另一大用途就是控制程序的运行顺序。

**Reply by whchrome** [reply]
明白了，谢谢

**Comment by tetsuoyyx** [comment]
阮老师您好，如果耗时VI出现了错误，会不会导致光标不能取消忙碌状态？

**Reply by ruanqizhen** [reply]
如果程序逻辑正确，是不会影响光标的。只要不是故意的，程序一出错，就不调用“消忙碌状态”就没问题。

**Comment by tetsuoyyx** [comment]
阮老师，我把布尔控件的引用传递给子VI处理，用属性节点访问引用指向对象的值，返回/写入的都是变体类型的？我想让传递的引用附带类型信息，这个可以做到吗？如果不行，我需要对变体做转换再转换回去吗？

**Reply by ruanqizhen** [reply]
如果子 VI 使用的是具体类型的引用，那么子 VI 中应该可以直接获得原来控件的数据类型，也就是直接使用布尔类型，而不是通用的变体类型。可以参考一下这里的例子： <https://lv.qizhen.xyz/vi_server_for_ui#通过引用调用属性节点>

**Comment by reinit258** [comment]
阮老师您好，我在子vi中用引用的方式使用了事件结构，在主vi中同样使用了事件结构，子vi事件能够正常运行，而主vi触发事件就会卡死。当然，我们应该尽量避免一个vi里出现多个事件结构，但是为了实现封装我貌似又不得不采用这样的结构，您有什么解决办法吗。附简图如下：![子vi中通过传引用使用事件结构](https://github.com/reinit258/source/blob/main/image.png?raw=true)

**Reply by ruanqizhen** [reply]
你的程序实质上是嵌套的事件结构，就是子 VI 的事件结构是在主 VI 的事件结构中某个分支里的。嵌套的事件结构本身就存在逻辑问题：程序要恰好按顺序出现相应的两个事件，它才有可能会运行到内层的事件结构，其他情况都可能造成程序死锁。
所以一定不要使用嵌套的事件结构。
最通常的解决办法是在主 VI 中使用一个事件结构捕获所有事件。子 VI 只用于实现某个事件需要调用的功能，而不是用于捕获该事件。

**Reply by reinit258** [reply]
感谢，一开始这个子vi我是放在外部调用一次的，后来看您文章里的事件驱动的结构，然后模仿了一下，把子vi放入用户定义事件里的初始化条件里了，后面程序也没补其他内容，可能就导致始终在初始化条件内了？。听您一说把子vi放到事件结构外果然解决了程序卡死的问题，但是我现在这样的结构是否还存在问题呢？我的目标是在子程序内实现一个列表框，通过选中实现项符号的变化，貌似目前效果处于预期。
![image](https://github.com/ruanqizhen/labview_book/assets/169650744/6bf2ac89-860c-43b7-ba29-a47933e6cf57)
![image](https://github.com/ruanqizhen/labview_book/assets/169650744/cb6fcc5c-3f55-4585-a5d6-d711428b741f)

**Reply by ruanqizhen** [reply]
只是简单的把子 vi 放在外部调用也不是特别好，相当于两个并行的事件结构，很容易搞不清楚哪个事件在哪个结构中被处理。子 VI 中是可以有使用件结构，不过需要遵循一些规范才不容易出错：
* 一个 VI 的所有事件都应该放在事件结构中进行处理，可以在主 VI 中，也可以才子 VI 中。就是说如果决定在子 VI 中处理所有事件，那么就不要在主 VI 中再使用事件结构了。
* 有的程序是有多个界面的。主程序中的事件结构处理所有主 VI 界面的事件；子 VI 有自己的界面，子 VI 中的事件结构处理所有子 VI 的界面事件。这样，两个事件结构互不干扰，也是可以的。

**Reply by reinit258** [reply]
好的，那现在看来您文章里使用到的该结构应当就是要将所有事件结构放到子vi中去了
![image](https://github.com/ruanqizhen/labview_book/assets/169650744/44674fd8-cac7-4655-a4c5-0b4311d884d3)
我仅仅将他作为一种封装某个功能的办法去使用，显然是错误的。
另外，您文章中提到的事件驱动的结构是否有案例呢，我没能搜索到。
![image](https://github.com/ruanqizhen/labview_book/assets/169650744/ffdcb764-9e6d-4d9a-8fc5-c19cfd6c6adf)
感谢老师的回答

**Reply by ruanqizhen** [reply]
相关的示例在这个文件夹里：  <https://github.com/ruanqizhen/labview_book/tree/main/code/Common_Patterns/事件结构和程序界面/>

"一般界面程序的框图.vi"

**Comment by ruanqizhen** [comment]
之前博客上的相关文章，搬到这里：

    Event Structure 也是一种选择结构，程序根据发生的事件决定执行哪一个页面的程序。此时，LabVIEW 的界面编写与 Visual Basic 的界面程序有些类似。

### 一. 按照产出源来区分事件的种类

    按照事件的产生源来区分，LabVIEW有以下几种事件：

![](http://tk1.storage.msn.com/x1pxOYwqu4SjF5G0W4dmEwaKLtSa4ws0-_l23pai0BiY4CUPEpY38Tn57aip6V5Fwz-WEpakzp6ekAKGVhg5rjFYhH72stkal4CjYJOK5mRJRo0w2rO5MMeUwrPJWbR-Bf_-Fj_GRR5deJHQLwztWU5h_cvhqVToeKm)  
图1：配置事件

    1． 应用程序事件（`<Application>`），这类事件主要反映整个应用程序状态的变化，例如：程序是否关闭，是否超时等。  
    2． VI事件（`<This VI>`），这类事件反映当前VI状态的改变。例如：当前VI是否被关闭，是否选择了菜单中的某一项等等。  
    3． 动态事件（Dynamic），用于处理用户自己定义的或在程序中临时生成的事件。  
    4． 区域事件（Pane）和分割线事件（Splitter）是LabVIEW 8中新添加的特性。LabVIEW 8中，用户可以把一个VI的前面板分割成几份，这两类事件用来处理用户对某个区域或区域分割线的状态的改变。

![](http://tk1.storage.msn.com/x1pxOYwqu4SjF5G0W4dmEwaKLtSa4ws0-_l23pai0BiY4CR364zchBRaD_O56J7B0GXsq4NAs94zXvXsLgLORFnygmC32Jb1cFfyUpegeHurPPjV8xUKnWRDqZy26BjwOfTdoLCF1J1SNL_gpBmn7l7VKubXRf2FTai)  
图2：面板上划分区域

    5． 控件事件（Control）是最常用的一种事件，用于处理某个控件状态的改变。例如，控件值的改变，或者鼠标键盘的操作。  
打开上述的“edit events”框，只要选定了某一个事件产生源，其相应的所有事件均排列在右侧events框中。  
有时候，多个事件产生源会对同一个用户操作分别产生相应事件。比如在某一控件上按下鼠标，区域事件和控件事件都会发出鼠标按下（Mouse Down）事件。LabVIEW 按以下规则顺序产生不同的事件：  
 键盘相关的事件（Key Down, Key Up, etc.）只在当前选中（Key Focused）的控件上产生;  
 鼠标相关的事件（Mouse Down, etc.）按照从外向里的顺序发出。例如，区域的鼠标按下事件先于控件的鼠标按下事件发出；结构的鼠标按下事件先于先于结构内控件的鼠标按下事件发出。  
 值改变事件按照从内向外的顺序发出。结构（Cluser）内控件的值改变事件先于结构的值改变事件发出。

### 二. 按照发出时间区分事件的种类

    按照事件的发出时间来区分，LabVIEW的事件可分为通知型事件（Notify Event）和过滤型事件（Filter Event）。

    通知型事件是在LabVIEW处理完用户操作之后发出的，比如用户利用键盘操作改变了一个字符串，LabVIEW 在改变了该控件的值之后，发出一个值改变（Value Changed）通知型事件，告诉事件结构，控件的值被改变了。如果事件结构内有处理该事件的框架，则程序转去执行该框架。  
    过滤型事件是在 LabVIEW 处理用户操作之前发出的，并等待相对应的事件框架执行完成之后，LabVIEW 再处理该用户操作。这类事件的名称之后都有一个问号。例如键盘按下？事件（Key Down? Event），当用户处理该事件时，控件的值还没有被改变，因此，用户可以在该事件对应的事件框架内决定是否让 LabVIEW先处理该事件，或改变键盘按下的值之后再让LabVIEW继续处理该事件。  
    可以明显地看出，过滤型事件比相应的通知型事件要先发出。  
    当同一VI的程序框图上有多于一个的事件结构时，通知型事件是同时被发往所有的事件结构的，而过滤型事件则是按顺序、依次发往每一个事件结构的。但是，在同一VI上放置多个事件结构是没有必要，而且极易引起错误的。所以应该避免在同一VI上使用多个事件结构。

    下面举例说明如何使用通知型事件。我们经常需要使用到这样的字符串控件：控件用于输入电话号码，因此只接收数字和横线，对其他按键不起反应。LabVIEW没有直接提供此种控件，但是它们可以利用通知型事件被方便地实现出来。

![](http://tk1.storage.msn.com/x1pxOYwqu4SjF5G0W4dmEwaKLtSa4ws0-_l23pai0BiY4DcR0SvOgQgxy0kEXYYEtJ_utYxLwrcoNqFs1Xc77XRR-UZGDNzUcDNu3HF4dfaflAdnsZw03kfI5xc9Sknx90z7zdiNHMHBVyh-s1eug3HbRZhRcUnIUVg)  
图3：利用 Key Down? 事件实现电话号码控件

    这个例子可以在这里下载：[](http://community.ni.com/examples/4f7f75288fc76ee4578b4e8b4ef65b9e73b076f463a553d765705b5776845b577b264e3263a74ef6/phone-number-control.vi)[Phone Number Control.vi](http://decibel.ni.com/content/docs/DOC-1076)

### 三. 动态事件

    在初始状态下，打开事件配置（Edit Events）对话框，动态事件下的一栏是空的。因为动态事件只有注册过之后才能使用。与事件相关的操作在函数选栏的 `Programming -> Dialog & User Interface -> Events` 下面。  
    用于注册事件的节点是事件注册节点（Register For Events）。需要注册某一事件时，先为它的产生者生成一个引用节点，然后将引用节点与事件注册节点的下方区域相连，再选取所需的事件。如下例：

![](http://tk1.storage.msn.com/x1pxOYwqu4SjF5G0W4dmEwaKLtSa4ws0-_l23pai0BiY4A8q2LirLQ2tyHqHa5EFiGJmCCf55d476NAKYG-TT9kDT6lW2QxkA9cmaxOKUNMBFk_dX594JvlNHRhZqw4SggxquAyFyRpLgrLwIme2BtcV_N-k5r_V2ai)  
图4：注册动态事件

    对于当前VI上的控件或区域等类型的事件，一般来说可以在事件结构中静态地被选择，所以不需要再动态注册一遍。但有时，当前VI的程序框图已经过于复杂，我们希望在子 VI 里去处理某些控件的事件。这时就可以把控件的引用传入子 VI，在子 VI 中动态注册所需事件。在子 VI 的事件结构中处理相应的事件。

### 四. 用户自定义的事件

    用户自定义的事件是动态事件的一种。用户自定义的事件不基于任何一个LabVIEW对象，它是使用创建用户事件节点（Create User Event）生成出来的。并且，用户可以选择不同的事件数据类型。

### 五. ActiveX控件的事件

    ActiveX 控件的事件不能直接被LabVIEW的事件结构所截获。ActiveX 事件需要用注册事件回调 VI 节点（Register Event Callback）来为某一事件指定一个VI。当事件发生时，执行被注册的 VI。我们也可以利用注册事件回调 VI 节点为某一 LabVIEW 自身的事件注册一个回调 VI，但是出于运行效率，和程序可读性等方面的考虑，最好不要这样使用。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

Unknown
2008年10月2日 22:32 编辑
請問 7.4版要做到事件觸發怎麼做?有可能做嗎?

我現在需要進行一個程序
點一下按鈕然後進行一個附程式並記錄數據
可以指導一下嗎?

回复
Qizhen
2008年09月2日 08:55 编辑
更新了下载链接

回复
ttongg
2008年08月29日 09:29 编辑
范例不能下载,能发我邮箱里一份吗?
谢谢!

tongxiaozhao@126.com

回复
丽芸
2007年10月10日 17:04 编辑
我是一个LV的初学者,最近一直在看你的博,呵呵,很多不懂,以后请多指教:>

回复
玉文
2007年09月3日 21:37 编辑
…..oooO…………………..(……)…Oooo………….)…/. …(…..)…………..(/…….)…/……………………. (/……………. Pasé por.. …………………. Aquí ………………oooO…………………..(……)…Oooo………….)…/. …(…..)……………(/…….)…/…………………….. (/……………………………………..
很喜欢你的博客，希望以后可以看到更多好咚咚，祝开心哦！

回复
Qizhen
2007年01月26日 18:37 编辑
问题是这样的，我手上没有 LabVIEW 7.1 啊。

回复
jiyang
2006年12月11日 10:24 编辑
大x侠的文章对我们来说非常有用，可是我现在还在用labVIEW7.1，不能打开你给的例子，不知道以后能否也提供一下这个版本的例子：）

回复
波
2006年12月9日 14:07 编辑
我是个新手，看了你的blog真是受益匪浅。很希望能结识前辈:(

**Comment by leilei319yj** [comment]
您好，
我从github上面下载了示例代码，为什么我使用Labview2025英文社区版打不开呢，打开的时候也没有报错，这是什么问题呢？谢谢

**Reply by ruanqizhen** [reply]
是任何一个示例都打不开吗？还是有一些打不开？
有可能现在 LabVIEW 无法打开太早之前版本的 VI 了。

**Reply by leilei319yj** [reply]
是都打不开

**Reply by ruanqizhen** [reply]
我现在电脑上安装的还是2021版的。等我找个时间，装一下2025，然后试试。

**Reply by leilei319yj** [reply]
好的，感谢

**Reply by ruanqizhen** [reply]
我今天安装了 2025 社区版，打开 VI 没有任何问题啊。
我是先打开 LabVIEW，然后选择打开 VI。
你是怎么打开的？

**Reply by leilei319yj** [reply]
我都试过，就是不行，先打开Labview然后选择打开VI的话啥也没有，也不报错，我再重装一下我的这个社区版。
非常感谢！

**Reply by ruanqizhen** [reply]
有可能的话，换个电脑试一下，也许跟系统有关系。

**Reply by leilei319yj** [reply]
好的，感谢

**Reply by ruanqizhen** [reply]
又想到一个可能的问题是系统的语言编码。你可以试试看把中文文件名换成纯英文，再试试能不能打开。

**Reply by leilei319yj** [reply]
真的是这个原因，太感谢了！
我的电脑是英文系统，我把文件改成一个英文名字然后放到一个完全英文的路径里面，就可以打开了。没有您的代码是痛并快乐着，我得花不少时间来才能整出来您的例子，整不出的时候特别郁闷，整出来了又特别高兴。

**Reply by ruanqizhen** [reply]
如果你的电脑不需要打开除了英语之外，其它国家语言编码的程序，那么可以把系统的默认语言编码改为中文 GB18030 编码，这样就能打开中文名的程序了。

这是 LabVIEW 一个比较讨厌的问题，书里有一段专门讨论了这个问题：<https://lv.qizhen.xyz/appendix_problem#labview-的-unicode-问题>

**Reply by leilei319yj** [reply]
好的，谢谢
